typescript/audio-summarizer/README.md

```diff
@@ -0,0 +1,188 @@
+
+In the `src/app/api/transcribe/route.ts` I will then create a [thread](https://docs.getliteral.ai/concepts/observability/thread) for each run. This is a bit of a hack as the interaction is not really a threaded conversation, however it is necessary so that I can upload Audio files.
+
+```ts
+// We receive the file from the frontend as a Blob, however Literal expects a ReadableStream so we have to convert it
+const nodeStream = Readable.fromWeb(formAudio.stream() as ReadableStream<any>);
+
+// Upload the file to Literal and add it as an attachment
+const { objectKey } = await literalClient.api.uploadFile({
+  content: nodeStream,
+  threadId: thread.id,
+  mime: "audio/webm",
+});
+const attachment = new Attachment({
+  name: "Audio file",
+  objectKey,
+  mime: "audio/webm",
+});
+
+// Create the run with the attached audio file
+const run = await thread
+  .step({
+    id: runId,
+    type: "run",
+    name: "Emojifier 🤖🎨",
+    input: {
+      input: { content: "Audio file" },
+      attachments: [attachment],
+    },
+  })
+  .send();
+```
+
+### Logging the steps
+
+Still in `src/app/api/transcribe/route.ts`, I then add the first step for the audio transcription. Please note that I am measuring start and end time, which will allow me to monitor latency from Literal AI.
+
+```ts
+await run
+  .step({
+    type: "llm",
+    name: "Audio transcription",
+    input: { content: "Audio file" },
+    output: { content: transcribedText },
+    attachments: [attachment],
+    startTime: start.toISOString(),
+    endTime: end.toISOString(),
+    generation: {
+      provider: "openai",
+      model: "whisper-1",
+      prompt: "See attached audio file",
+      completion: transcribedText,
+    },
+  })
+  .send();
+```
+
+Next in `src/app/api/emojify/route.ts`, I need to fetch the run and add the second step for the summarization. To capture the run time of the LLM call and a host of other informations such as token counts, I will use the built-in OpenAI instrumentation exposed by the Literal AI SDK.
+
+```ts
+// Fetch the run
+const runData = await literalClient.api.getStep(runId);
+if (!runData) {
+  return new NextResponse("Run not found", { status: 404 });
+}
+// This step will just instantiate the run data as a new Step instance so it can be used later
+const run = literalClient.step(runData);
+
+// Call the LLM API
+const completion = await openai.chat.completions.create({
+  ...prompt.settings,
+  messages: [...promptMessages, { role: "user", content: text }],
+});
+// Instrument the call to OpenAI using Literal AI SDK
+await literalClient.instrumentation.openai(completion, run);
+```
+
+Lastly, I will patch the Run by providing it the end time and the completion data. This will allow me to monitor the overall latency of each run, including the network latency from one call to the other.
+
```

Comment:
Still surprised that you have to provide the times yourself. I can see latency numbers on Literal AI steps.
![Uploading Screenshot from 2024-06-18 12-47-32.png…]()

Reply:
In the case of a Run it is not automatic, and adding it manually allows you to get the "real" latency (including network latencies etc...) which is higher than just summing up the latency of each LLM call. I've seen a cookbook somewhere where it was done like that (i can't seem to find it again but uuuh... trust me bro !)